### ISSUE TITLE
Add Ability to Replace Pencil & Private Stencil Collections

### ISSUE THREAD

**Issue by prikhi** [prikhi/pencil#678]
Currently users if users want to install a newer version of a collection, they must remove the old one first. Included collections cannot be removed so they are hardcoded to whichever version was included . This makes bundled 3rd party collections un-upgradeable. We should allow replacing anything...

We could simply always ask the user if they wished to replace the existing collection, then delete the collection before continuing with the import or abort the import.

Or we could add something like version numbers to collections & allow replacing if the version number is higher.

See: [collectionManager.js](https://github.com/prikhi/pencil/blob/develop/app/content/pencil/definition/collectionManager.js#L250-L254)

**Comment by zenware** [comment]
I'm not too familiar with Pencil Project, but I just read through #490 and I was wondering when you are talking about "Bundled Collections" are you referring to a collection that you install all at once as a group, or the collection that distributes with pencil itself? Because if it's the former, it wouldn't be too difficult to build a dependency management solution.

**Comment by prikhi** [comment]
> talking about "Bundled Collections" are you referring to a collection that you install all at once as a group, or the collection that distributes with pencil itself

By `Bundled Collections`, I meant 3rd-party collections that would eventually be included by #677. 

In general this ticket was meant for both(I believe Pencil treats them equally under the hood but I haven't completely immersed myself in the stencil collection code yet). But now that I think about it, the latter isn't as big a deal - I'm failing to come up with reasonable circumstances where someone might want to replace a built-in collection instead of just upgrading(maybe an incompatible version of FF/XR?).

> it wouldn't be too difficult to build a dependency management solution

By dependency management, do you mean version management? As far as I know there aren't dependencies between stencil collections(where one collection requires that another is installed).

I was thinking stencil collections could have a `version`/`last-updated`/`modified` attribute or something similar, and Pencil could ensure you only replace a collection with a collection that has a higher version number. I don't see versioning as a _necessary_ requirement for this ticket, we could just let users replace a collection with any version of that collection.

**Comment by zenware** [comment]
Yes, I definitely meant _version management_, that is entirely my mistake.
If we base it on github repositories it can easily check if there is an update to the repository and grab it automatically. But as I mentioned in #677 I don't think it should be with submodules. Instead it should work like other github based package managers and just take a `username/repository` and clone them all into the correct place. Then it would be as simple as running a `git pull` on each collection to update when the application is launching.

**Comment by prikhi** [comment]
Sorry I misspoke as well, I meant to say this ticket was meant for _all_ collections - built-in collections, bundled collections, **and** user-created/installed collections.

> If we base it on github repositories it can easily check if there is an update to the repository and grab it automatically.
> ...
> Then it would be as simple as running a git pull on each collection to update when the application is launching.

While I'm not opposed to automatically updating stencil collections, I meant for this issue to be more about allowing end-users to manually replace them on their individual installation/profile instead of requiring the stencil to be uninstalled first. So it seems to me to be a bit out of scope for this issue(correct me if I'm misunderstanding), but feel free to create a new issue for it if you'd like(or I can when I have some more free time).
